### PR TITLE
Cybersource Rest: Support NT

### PR DESCRIPTION
**Pull request by aenand**
COMP-75

Adds support for Network Tokens on the Cybersource Rest gateway.

Test Summary
Local:
5848 tests, 79228 assertions, 0 failures, 23 errors, 0 pendings, 0 omissions, 0 notifications 99.6067% passed
Errors unrelated to this gateway
Unit:
32 tests, 160 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications 100% passed
Remote:
50 tests, 157 assertions, 7 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications 86% passed
Errors also on master

### CODE REVIEW

File: test/unit/gateways/cyber_source_rest_test.rb [hunk 1]
```diff
@@ -187,6 +203,34 @@ def test_add_shipping_address
     assert_equal '4158880000', address[:phoneNumber]
   end
 
+  def test_authorize_network_token_visa
```
Comment (almalee24) [inline test/unit/gateways/cyber_source_rest_test.rb:206]: I know these tickets aren't together but could you please add tests for stored credentials just to ensure that we are properly sending all the fields.

File: lib/active_merchant/billing/gateways/cyber_source_rest.rb [hunk 2]
```diff
@@ -191,25 +196,31 @@ def add_payment(post, payment, options)
...
       def add_network_tokenization_card(post, payment, options)
-        post[:processingInformation][:paymentSolution] = PAYMENT_SOLUTION[payment.source]
         post[:processingInformation][:commerceIndicator] = 'internet' unless card_brand(payment) == 'jcb'
 
         post[:paymentInformation][:tokenizedCard] = {
           number: payment.number,
           expirationMonth: payment.month,
           expirationYear: payment.year,
           cryptogram: payment.payment_cryptogram,
-          transactionType: '1',
```
Comment (almalee24) [inline lib/active_merchant/billing/gateways/cyber_source_rest.rb:202]: I think ApplePay/GooglePay still need a transactionType of 1

Comment (almalee24) [inline lib/active_merchant/billing/gateways/cyber_source_rest.rb:202]: Why not just keep it here an do something like payment_method.source == :network_token ? '3' : '1'?

File: lib/active_merchant/billing/gateways/cyber_source_rest.rb [hunk 3]
```diff
@@ -191,25 +196,31 @@ def add_payment(post, payment, options)
       end
 
       def add_network_tokenization_card(post, payment, options)
-        post[:processingInformation][:paymentSolution] = PAYMENT_SOLUTION[payment.source]
         post[:processingInformation][:commerceIndicator] = 'internet' unless card_brand(payment) == 'jcb'
```
Comment (almalee24) [inline lib/active_merchant/billing/gateways/cyber_source_rest.rb:199]: commerceIndicator needs to be updated base on stored_credentials

File: lib/active_merchant/billing/gateways/cyber_source_rest.rb [hunk 4]
```diff
@@ -191,25 +196,31 @@ def add_payment(post, payment, options)
...
-            ucafAuthenticationData: payment.payment_cryptogram,
-            ucafCollectionIndicator: '2'
-          }
+        if payment.source == :network_token
+          post[:paymentInformation][:tokenizedCard][:transactionType] = '3'
+          post[:processingInformation][:paymentSolution] = NT_PAYMENT_SOLUTION[payment.brand] if NT_PAYMENT_SOLUTION[payment.brand]
         else
-          post[:consumerAuthenticationInformation] = { cavv: payment.payment_cryptogram }
+          # Apple Pay / Google Pay
+          post[:processingInformation][:paymentSolution] = WALLET_PAYMENT_SOLUTION[payment.source] if WALLET_PAYMENT_SOLUTION[payment.source]
```
Comment (almalee24) [inline lib/active_merchant/billing/gateways/cyber_source_rest.rb:214]: Would it be safe to remove `if WALLET_PAYMENT_SOLUTION[payment.source]` considering that ApplePay or GooglePay are the only options

File: test/unit/gateways/cyber_source_rest_test.rb (hunk 1 above)
Comment (aenand) [reply test/unit/gateways/cyber_source_rest_test.rb:206]: There's another PR to update the stored credential mappings!

File: lib/active_merchant/billing/gateways/cyber_source_rest.rb (hunk 2 above)
Comment (aenand) [reply lib/active_merchant/billing/gateways/cyber_source_rest.rb:202]: We could do that!

File: lib/active_merchant/billing/gateways/cyber_source_rest.rb (hunk 3 above)
Comment (aenand) [reply lib/active_merchant/billing/gateways/cyber_source_rest.rb:199]: There's another PR that updates this value!

File: lib/active_merchant/billing/gateways/cyber_source_rest.rb (hunk 4 above)
Comment (aenand) [reply lib/active_merchant/billing/gateways/cyber_source_rest.rb:214]: 🤔 probably!

File: lib/active_merchant/billing/gateways/cyber_source_rest.rb [hunk 5]
```diff
@@ -191,25 +196,30 @@ def add_payment(post, payment, options)
...
-          post[:consumerAuthenticationInformation] = {
-            ucafAuthenticationData: payment.payment_cryptogram,
-            ucafCollectionIndicator: '2'
-          }
+        if payment.source == :network_token && NT_PAYMENT_SOLUTION[payment.brand]
+          post[:processingInformation][:paymentSolution] = NT_PAYMENT_SOLUTION[payment.brand]
         else
-          post[:consumerAuthenticationInformation] = { cavv: payment.payment_cryptogram }
+          # Apple Pay / Google Pay
+          post[:processingInformation][:paymentSolution] = WALLET_PAYMENT_SOLUTION[payment.source]
```
[the remark this answers is not in the document]
Comment (aenand) [reply lib/active_merchant/billing/gateways/cyber_source_rest.rb:214]: For rest, yes!

**Review by almalee24 — APPROVED** [review]

File: lib/active_merchant/billing/gateways/cyber_source_rest.rb [hunk 6]
```diff
@@ -93,6 +98,7 @@ def scrub(transcript)
           gsub(/(\\?"number\\?":\\?")\d+/, '\1[FILTERED]').
           gsub(/(\\?"routingNumber\\?":\\?")\d+/, '\1[FILTERED]').
           gsub(/(\\?"securityCode\\?":\\?")\d+/, '\1[FILTERED]').
+          gsub(/(\\?"cryptogram\\?":\\?")[^<]+/, '\1[FILTERED]').
```
Comment (aenand) [inline lib/active_merchant/billing/gateways/cyber_source_rest.rb:101]: @almalee24 also added scrubbing of cryptogram and a test for it.

Comment (almalee24) [reply lib/active_merchant/billing/gateways/cyber_source_rest.rb:101]: Got it, thanks for catching that!

**Pull request merged**